### PR TITLE
[536] [913] [910] SchemaMigrator/SchemaValidator support for PostgreSQL and CockroachDB

### PR DESCRIPTION
**Pull request by gbadner**
For https://github.com/hibernate/hibernate-reactive/issues/536, https://github.com/hibernate/hibernate-reactive/issues/913, and https://github.com/hibernate/hibernate-reactive/issues/910 .

Requires ORM PR:
https://github.com/hibernate/hibernate-orm/pull/4106

### CODE REVIEW

**Comment by DavideD** [comment]
I will convert this to a draft, because it contains changes to the `gradle.properties` file that we don't want on the main branch. And also because we have to wait for ORM to include the PR that we need.

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/SchemaUpdatePostgreSqlTestBase.java [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+																	 )
+												)
+										)
+										.thenCompose( v -> s.createNativeQuery( indexDefinitionQuery, String.class )
+												.setParameter( 1, "aanother" )
+												.getSingleResult()
+												.thenAccept( result ->
+																	 context.assertEquals(
+																			 result,
+																			 "CREATE UNIQUE INDEX aanother_pkey ON public.aanother USING btree (id)"
```
Comment (DavideD) [inline hibernate-reactive-core/src/test/java/org/hibernate/reactive/SchemaUpdatePostgreSqlTestBase.java:210]: Not critical, but when using `context.equals(expected, actual)`, the firs value is the expected one and the second the actual one.

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/pool/ReactiveConnection.java [hunk 2]
```diff
@@ -50,6 +50,7 @@
 	CompletionStage<Result> select(String sql);
 	CompletionStage<Result> select(String sql, Object[] paramValues);
 	CompletionStage<ResultSet> selectJdbc(String sql, Object[] paramValues);
+	CompletionStage<ResultSet> selectJdbcOutsideTransaction(String sql, Object[] paramValues);
```
Comment (DavideD) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/pool/ReactiveConnection.java:53]: Could you remind us why you need the additional method?

**Comment by gavinking** [comment]
I have not tested it but the code looks great! Thanks @gbadner 

(The only very minor comment I have is I would prefer that SQL keywords were in lowercase like the rest of the SQL we generate.)

**Comment by gbadner** [comment]
@DavideD, thanks for doing making it a draft. I thought I defined it as a draft when I created it, but maybe I missed that.

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/SchemaUpdatePostgreSqlTestBase.java (hunk 1 above)
Comment (gbadner) [reply hibernate-reactive-core/src/test/java/org/hibernate/reactive/SchemaUpdatePostgreSqlTestBase.java:210]: I normally do it that way, I don't know why I didn't in this test. I'll change it.

Comment (gbadner) [reply hibernate-reactive-core/src/test/java/org/hibernate/reactive/SchemaUpdatePostgreSqlTestBase.java]: @DavideD, I pushed a [commit|https://github.com/hibernate/hibernate-reactive/pull/903/commits/99eb01a207506b1536aee6ca7ed4a2252db57632] that corrects the order.

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/pool/ReactiveConnection.java (hunk 2 above)
Comment (gbadner) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/pool/ReactiveConnection.java:53]: @DavideD , since `SchemaMigrator` executes select queries on system tables and executes DDL statements (which update system tables), I assumed that these queries should be executed outside of a transaction (i.e., with `autocommit=true`). IIUC, this would help avoid contention with other work the DB is doing on schemas used by various applications.

If you look at the corresponding ORM code in [GenerationTargetToDatabase]
(https://github.com/hibernate/hibernate-orm/blob/main/hibernate-core/src/main/java/org/hibernate/tool/schema/internal/exec/GenerationTargetToDatabase.java#L74-L85): you can see that the statement is created using a connection obtained from a `DdlTransactionIsolator`.

Javadoc for `DdlTransactionIsolator` says:
```
 * Provides access to a Connection that is isolated from
 * any "current transaction" with the designed purpose of
 * performing DDL commands
```
If you look at the `DdlTransactionIsolator` implementations, you will see that `autocommit` is set to `true` on the `Connection` that is obtained:
`DdlTransactionIsolatorJtaImpl`
`DdlTransactionIsolatorNonJtaImpl`
`DdlTransactionIsolatorProvidedConnectionImpl` relies on `JdbcConnectionAccessProvidedConnectionImpl` which sets `autocommit` to `true`

I didn't see any other way to execute a query outside of a transaction that returned a `CompletionStage<ResultSet>`, so I created a new method.

Is there some other method that should be used instead?

**Comment by gbadner** [comment]
> I have not tested it but the code looks great! Thanks @gbadner

Thanks @gavinking!

> 
> (The only very minor comment I have is I would prefer that SQL keywords were in lowercase like the rest of the SQL we generate.)

That's fine. I can change that.

**Comment by gbadner** [comment]
I'm currently using the Hibernate ORM codestyle. Which of the [codestyles](https://github.com/hibernate/hibernate-ide-codestyles/tree/main/intellij-14) do you use for Hibernate Reactive?

**Comment by gbadner** [comment]
I finally got around to taking a good look at what else needs to be done for `SchemaValidator`, and, AFAICT, it's already working with the code provided here!?!

**Comment by gbadner** [comment]
@gavinking, I added a new [commit](https://github.com/hibernate/hibernate-reactive/pull/903/commits/ba2e5f9411c1af09e61c65e6ad7946f8e9aeae3f) that changes SQL keywords to lower-case.

**Comment by gavinking** [comment]
Thanks, @gbadner

**Comment by DavideD** [comment]
> @DavideD, thanks for doing making it a draft. I thought I defined it as a draft when I created it, but maybe I missed that.

no problem

> I'm currently using the Hibernate ORM codestyle. Which of the [codestyles](https://github.com/hibernate/hibernate-ide-codestyles/tree/main/intellij-14) do you use for Hibernate Reactive?

Yes, that's what we should use. Thanks.

Although, our build doesn't check the code style (except of a couple of basic rules) so you might find that we are a bit inconsistent sometimes.

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/pool/ReactiveConnection.java (hunk 2 above)
Comment (DavideD) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/pool/ReactiveConnection.java:53]: I don't think we had this requirement before. I don't see any problem adding a new method.

But I mainly wanted a bit of an explanation written somewhere so that others know as well (maybe it wasn't obvious only to me though ;-)

At some point, we will need to add some javadoc to that class. Anyway, not something you need to worry about at the moment.

Comment (gbadner) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/pool/ReactiveConnection.java:53]: Yes, I know this stuff is lacking in documentation. I will be adding that.

**Comment by gbadner** [comment]
s@DavideD, I just pushed a commit that adds Javadoc and a couple of small tweaks.

I forgot that I had already pushed a commit for encapsulating methods into `ExtractionTool` and ended up squashing the two commits, so I had to force-push.

Sanne is still looking at the ORM PR. I think this PR is ready to review, although we will obviously have to wait for the ORM PR to be finalized.

Unless I hear of any required changes to this PR, I will start work on adding support for MySQL.

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/adaptor/impl/ResultSetAdaptor.java [hunk 3]
```diff
@@ -430,6 +430,8 @@ public String getCatalogName(int column) {
 
 			@Override
 			public String getColumnTypeName(int column) {
+				// This information is in rows.columnDescriptors().get( column-1 ).dataType.name
+				// but does not appear to be accessible.
```
Comment (gavinking) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/adaptor/impl/ResultSetAdaptor.java:434]: @gbadner this is used by `SchemaValidator`, is that right?

Should we ask the vert.x guys to expose this info on `ColumnDescriptor`?

**Comment by gavinking** [comment]
This all looks very nice, @gbadner!

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/adaptor/impl/ResultSetAdaptor.java (hunk 3 above)
Comment (gbadner) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/adaptor/impl/ResultSetAdaptor.java:434]: @gavinking,

TL; DR: yes, it would help to expose this info on `ColumnDescriptor`.

For the record, here are the details:

If they expose it, then we can use `ResultSetMetaData` like what is done in ORM [here](https://github.com/hibernate/hibernate-orm/blob/a64d0102812474898059f1143bf09764f95556f1/hibernate-core/src/main/java/org/hibernate/tool/schema/extract/internal/InformationExtractorJdbcDatabaseMetaDataImpl.java#L152-L167) in the ORM [PR](https://github.com/hibernate/hibernate-orm/pull/4106).

Currently, since that information is not available in `ResultSetMetaData`, [`AbstractReactiveInformationSchemaBasedExtractorImpl#processColumnsResultSet`](https://github.com/hibernate/hibernate-reactive/pull/903/files#diff-149c9cc5fe97eafff78515120d487c5934371492bd13929253d0ec88278e9316R220-R259) is executed instead. 

Notice also that the SQL type code gets returned as `null` by `AbstractReactiveInformationSchemaBasedExtractorImpl#processColumnsResultSet` (which ends up being 0 (zero), which is java.sql.Types.NULL) because it is not available. It would be helpful to get that SQL code; otherwise, this [line](https://github.com/hibernate/hibernate-orm/blob/main/hibernate-core/src/main/java/org/hibernate/tool/schema/internal/AbstractSchemaValidator.java#L165) constructing an error message might not make sense (or may throw an exception?).

From looking at the code, it appears that `ResultSetAdapter`'s implementation of `ResultSetMetaData#getColumnType` returns the SQL type code properly, but I haven't debugged into it to be sure.

AFAICT, `ColumnInformation#getNullable`, `#getColumnSize`, and `#getDecimalDigits` are not used by ORM's main branch, so if `ColumnDescriptor#dataType#name` gets exposed, we'll have everything ORM currently needs. I know that there has been talk about using some of this information in the future (in ORM 6?), but I'm not sure about that.

**Comment by gbadner** [comment]
Thanks @gavinking, I'm pretty happy with it.

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/provider/service/ReactiveImprovedExtractionContextImpl.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+		service = registry.getService( ReactiveConnectionPool.class );
+	}
+
+	@Override
+	public <T> T getQueryResults(
+			String queryString,
+			Object[] positionalParameters,
+			ResultSetProcessor<T> resultSetProcessor) throws SQLException {
+
+		final CompletionStage<ReactiveConnection> connectionStage = service.getConnection();
```
Comment (gbadner) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/provider/service/ReactiveImprovedExtractionContextImpl.java:69]: @gavinking, I noticed that [`ReactiveGenerationTarget#executeCommands`](https://github.com/hibernate/hibernate-reactive/blob/main/hibernate-reactive-core/src/main/java/org/hibernate/reactive/provider/service/ReactiveGenerationTarget.java#L91-L105) uses the same `ReactiveConnection` to execute all of the DDL statements.

Here, `Service#getConnection` gets called for each query that gets executed. That `ReactiveConnection` gets closed in the `finally` block below.

Would it be better to keep a `ReactiveConnection` around so the same one is used for all of the queries?

Comment (gavinking) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/provider/service/ReactiveImprovedExtractionContextImpl.java:69]: I can't think of a reason why it would matter.

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/adaptor/impl/ResultSetAdaptor.java (hunk 3 above)
Comment (gavinking) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/adaptor/impl/ResultSetAdaptor.java:434]: I opened https://github.com/eclipse-vertx/vertx-sql-client/issues/1020.

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/provider/service/ReactiveImprovedExtractionContextImpl.java (hunk 4 above)
Comment (gbadner) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/provider/service/ReactiveImprovedExtractionContextImpl.java:69]: OK, I'll leave it as is.

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/provider/service/AbstractReactiveInformationSchemaBasedExtractorImpl.java [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+			Object parameter,
+			StringBuilder sb,
+			List<Object> parameters) {
+
+		if ( parameter != null ) {
+			parameters.add( parameter );
+			sb.append( clause );
+			// TODO: PostgreSQL does not seem to accept "?" parameter placeholders.
+			//       It does accept $<position>. Will this work for all dialects?
+			sb.append( "$").append( parameters.size() );
```
Comment (gbadner) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/provider/service/AbstractReactiveInformationSchemaBasedExtractorImpl.java:171]: I see that this doesn't work for MySQL. It looks like `PostgresParameters` knows to make this change.

I'm looking into how I can change this to use `Parameters`...

Comment (gbadner) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/provider/service/AbstractReactiveInformationSchemaBasedExtractorImpl.java]: I just added a [commit](https://github.com/hibernate/hibernate-reactive/pull/903/commits/cd964b9b5609045249954280779c117e85d2d452) to use `Parameters` to get the dialect-specific placeholder.

**Comment by gbadner** [comment]
I just added a [commit](https://github.com/hibernate/hibernate-reactive/pull/903/commits/82fb199aa1e0b4e62f91c7a9704a079a455e9fb1) to deal with a non-standard information_schema.columns column name used by PostgreSQL.

**Comment by gbadner** [comment]
I pushed a [commit](https://github.com/hibernate/hibernate-reactive/pull/903/commits/63041ad00c0e90119b5df48f8cfcfdeb2e7920ea) that deals with PostgreSQL's storing sequence metadata as a String when ORM expects a Long.

This was the issue I mentioned [here](https://github.com/hibernate/hibernate-reactive/issues/536#issuecomment-848386513).

File: hibernate-reactive-core/src/main/java/org/hibernate/reactive/adaptor/impl/ResultSetAdaptor.java [hunk 6]
```diff
@@ -214,7 +214,20 @@ public int getInt(String columnLabel) {
...
+		Long aLong;
+		try {
+			aLong = row.getLong(columnLabel);
+		}
+		catch (ClassCastException ex) {
+			// Check if the value is a String that can be converted to a Long
+			final String aString = row.getString(columnLabel);
+			// aString won't be null; check just because...
+			aLong = aString != null ? Long.parseLong(aString) : null;
+		}
```
Comment (gavinking) [inline hibernate-reactive-core/src/main/java/org/hibernate/reactive/adaptor/impl/ResultSetAdaptor.java:229]: @gbadner won't this swallow the original CCE and cause it to be rethrown as something potentially misleading if it's *not* a string?

Comment (gbadner) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/adaptor/impl/ResultSetAdaptor.java:229]: Yes, you're right! I'll fix it to throw the original CCE in that case.

Comment (gavinking) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/adaptor/impl/ResultSetAdaptor.java:229]: Perfect, thanks 🙏

Comment (gbadner) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/adaptor/impl/ResultSetAdaptor.java:237]: @gavinking, I pushed a [commit](https://github.com/hibernate/hibernate-reactive/pull/903/commits/a4ec4b3e14b75639acbef9c8ff20a384534f20e6) to fix this. Does it look OK to you?

Comment (gavinking) [reply hibernate-reactive-core/src/main/java/org/hibernate/reactive/adaptor/impl/ResultSetAdaptor.java:237]: Yup that's fine.

**Comment by gbadner** [comment]
Pushed a [commit](https://github.com/hibernate/hibernate-reactive/pull/903/commits/f0639b1d1a5088be83965c18c417000e19cfe68a) for fixing on CockroachDB.

**Comment by gbadner** [comment]
There are small differences in results returned from querying the `pg_indexes` system table using CockroachDB compared to PostgreSQL. This only affected the CockroachDB unit test.

It would be better to change the test to not rely on `pg_indexes`. I think that would allow changing `SchemaUpdatePostgreSqlTestBase` to be used for both PostgreSQL and CockroachDB.

**Comment by gbadner** [comment]
@gavinking, rebased and settings in build.properties commented out.

**Comment by gbadner** [comment]
@gavinking,I'm seeing failures in tests that restrict ORM versions [5.5,5.6).

Should I remove .github/workflows/tracking-orm-5.build.yml ?

**Comment by gavinking** [comment]
Umm, I'm not sure TBH. Perhaps @Sanne knows what to do.

**Comment by gbadner** [comment]
@gavinking, I'd like to add the commits from https://github.com/hibernate/hibernate-reactive/pull/940 into this PR, then close https://github.com/hibernate/hibernate-reactive/pull/940.

Any reason why I shouldn't do that?

**Comment by gavinking** [comment]
Fine by me.

**Comment by gbadner** [comment]
Merged commits from https://github.com/hibernate/hibernate-reactive/pull/940.

**Comment by gavinking** [comment]
> Merged commits from #940.

Great.

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/BaseReactiveTest.java [hunk 7]
```diff
@@ -124,17 +126,17 @@ protected Configuration constructConfiguration() {
 			doneTablespace = true;
 		}
 		//Use JAVA_TOOL_OPTIONS='-Dhibernate.show_sql=true'
-		configuration.setProperty( Settings.SHOW_SQL, System.getProperty(Settings.SHOW_SQL, "false") );
+		configuration.setProperty( Settings.SHOW_SQL, System.getProperty(Settings.SHOW_SQL, "true") );
```
Comment (gavinking) [inline hibernate-reactive-core/src/test/java/org/hibernate/reactive/BaseReactiveTest.java:129]: @gbadner could you roll back this change please?

P.S. A more convenient way to enable SQL logging during development is to set:

    logger.hibernate.level = debug

in `log4j.properties`.

**Review by gavinking — CHANGES_REQUESTED** [review]

**Comment by gavinking** [comment]
> I'm seeing failures in tests that restrict ORM versions [5.5,5.6).

Is this the only thing blocking us from merging this?

**Comment by Sanne** [comment]
We should no longer test with Hibernate ORM 5.5 since this requires those changes which are exclusively in 5.6

**Comment by Sanne** [comment]
P.S. I've already upgraded Quarkus to use ORM `5.6.0.Beta1` so I'm not concerned about dropping compatibility with 5.5

**Comment by gavinking** [comment]
> We should no longer test with Hibernate ORM 5.5 since this requires those changes which are exclusively in 5.6

Yes, I agree.

**Comment by gavinking** [comment]
So am I understanding correctly that as of this change, we now have two redundant runs of the test suite for every CI build?

**Comment by gavinking** [comment]
Anyway, OK, right, @gbadner, this seemed to fix it:

<https://github.com/gavinking/hibernate-reactive/commit/7d8e19b88206e3b32697397ca804045762192e42>

**Comment by gbadner** [comment]
I pushed a commit that disables SQL logging and includes @gavinking 's  commit to fix DI. Crossing my fingers that everything passes...

**Comment by gbadner** [comment]
Hurray, all passed!

**Comment by gavinking** [comment]
WDYT shall we merge it?

**Comment by gbadner** [comment]
@gavinking, yes, please do!

**Comment by gavinking** [comment]
It's still a "Draft" :)

**Comment by gbadner** [comment]
Oops, it's "Open" now.

**Comment by gavinking** [comment]
There's a lot of commits here, should we squash?

File: hibernate-reactive-core/src/test/java/org/hibernate/reactive/BaseReactiveTest.java (hunk 7 above)
Comment (gbadner) [reply hibernate-reactive-core/src/test/java/org/hibernate/reactive/BaseReactiveTest.java]: @gavinking, this is done.

**Comment by gbadner** [comment]
Sure, I'll squash and push to my branch...

**Review by gavinking — APPROVED** [review]

**Comment by gavinking** [comment]
Excellent! Fantastic to finally have this done!

Thanks @gbadner

**Comment by Sanne** [comment]
Nice!
